server.go

```diff
@@ -3287,7 +3287,7 @@ func newServer(ctx context.Context, listenAddrs []string, db database.DB, chainP
 				return &rpcBlockTemplater{s.bg}
 			}(),
 			CPUMiner:             &rpcCPUMiner{s.cpuMiner},
-			TxIndex:              s.txIndex,
+			TxIndexer:            newRPCTxIndexer(s.txIndex),
```

Comment:
`rpcTxIndexer` and `rpcAddrIndexer` aren't actually defining any of the interface methods (they are just embedding the actual indexes, as you are suggesting).  The `newRPCTxIndexer` and `newRPCAddrIndexer` methods are just handling returning nil if the underlying index is nil, so that we don't end up with a non-nil interface wrapping a nil index.

Reply:
I see what you are saying now, to just get rid of the adaptors altogether, but the problem is you would still end up with a non-nil interface wrapping a nil pointer, but I think we could do something like this without the adaptor:
```
TxIndexer: func() rpcserver.TxIndexer {
	if s.txIndex == nil {
		return nil
	}
	return s.txIndex
}(),
```

Reply:
Updated to this and removed the adaptors.

Reply:
Right, except `s.txIndex` (and s.addrIndex`) are already concrete types that will already either be `nil` or not, so you don't even need the wrapper func.

Reply:
If you only put:
```
TxIndexer:            s.txIndex,
```
then `s.cfg.TxIndexer == nil` will still return `false` when `s.txIndex` is `nil`.  Since `s.cfg.TxIndexer` is an interface, `s.cfg.TxIndexer == nil` will only return true if both the dynamic type and value of the interface are `nil`, and when assigning `TxIndexer: s.txIndex` directly, the dynamic type of the interface will be `*indexers.TxIndex`.  The wrapper function assigns it as a `nil` dynamic type and value in the case that `s.txIndex` is `nil`. We could do it the way you are mentioning as well, but then we'd have to do a typecheck or use reflection instead of the `nil` checks that are currently in the handlers.

Reply:
Alright, I looked at the code in the rpcserver and I see it was changed a bit as to how I remembered it, so you're right.  You do indeed need the immediately-invoked func to explicitly return `nil`. 

An alternative is to create the config struct separately without any optional values specified (which will naturally make them be the zero value) and only set them in the case their implementer is not nil.

e.g.

```Go
rpcsConfig := rpcserver.Config{
	...
}
if s.bg != nil {
	rpcsConfig.BlockTemplater = &rpcBlockTemplater{s.bg}
}
if s.txIndex != nil {
	rpcsConfig.TxIndexer = s.txIndex
}
...

s.rpcServer, err = rpcserver.New(&rpcsConfig)
...
```

At any rate, the func approach is already used in various cases, so it probably makes sense to keep it consistent.